File: 剑指 Offer（第 2 版）/34.cpp
// 剑指 Offer 34. 二叉树中和为某一值的路径
// https://leetcode-cn.com/problems/er-cha-shu-zhong-he-wei-mou-yi-zhi-de-lu-jing-lcof/

#include<vector>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
private:
    // 当前路径
    vector<int> path;
    vector<vector<int>> res;
public:
    vector<vector<int>> pathSum(TreeNode* root, int target) {
        dfs(root, target);
        return res;
    }
    void dfs(TreeNode* curr, int target){
        if(curr){
            target -= curr->val;
            path.push_back(curr->val);
            if(target!=0 || curr->left || curr->right){
                dfs(curr->left, target);
                dfs(curr->right, target);
            }else{
                res.push_back(path);
            }
            // 回溯
            path.pop_back();
        }
    }
};

File: 剑指 Offer（第 2 版）/34.java
class Solution {
    List<List<Integer>> res = new ArrayList<>();
    LinkedList<Integer> path = new LinkedList<>();
    public List<List<Integer>> pathSum(TreeNode root, int target) {
        recur(root, target);
        return res;
    }
    public void recur(TreeNode root, int target){
        if(root==null) return;
        path.add(root.val);
        if(target==root.val && root.left==null && root.right==null) res.add(new LinkedList(path));
        recur(root.left, target-root.val); 
        recur(root.right, target-root.val); 
        path.removeLast();
    }
}

/*
注意事项
1. LinkedList removeLast
2. res.add的时候要新建一个对象，否则全是同一个list（path）
3. 不能通过判断当前val和target的大小关系进行剪枝，因为要考虑正负数问题
*/